GitCommitBear.py: Require URL in commit message body

nkprince007 opened the pull request:
This ensures that the git commit contains a URL
that relates to an issue.

Fixes https://github.com/coala/coala-bears/issues/1112

Mixih commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -197,6 +200,27 @@ def check_body(self, body,
+                path = pieces.path
+                try:
+                    assert all([scheme, netloc, path])
+                    if netloc.find('github') != -1 and path.find('issues'):
```
Not all people use github so we should have this configurable for those who use other solutions

meetmangukiya replied:
Also, we can't say if all the issue trackers will have *issues* in their URL

nkprince007 commented:
I'll update it to support more VCS. Thanks for reminding @Mixih

meetmangukiya commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -175,6 +177,8 @@ def check_imperative(self, paragraph):
+                   should_contain_url: bool=False,
                    ignore_length_regex: typed_list(str)=()):
         """
         Checks the given commit body.
```
Add the parameter doc strings for `vcs_used` and `should_contain_url`

nkprince007 replied:
Ok. @meetmangukiya

abhsag24 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -183,6 +187,9 @@ def check_body(self, body,
                                     newline character at each line end does not
                                     count to the length.
+        :param vcs_used:            The version control system used.
+        :param contains_url:  Whether the body contains a issue related
```
wrong indentation

nkprince007 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -197,6 +204,27 @@ def check_body(self, body,
+                path = pieces.path
+                try:
+                    assert all([scheme, netloc, path])
+                    if netloc.find(vcs_used) != -1 and path.find('issues'):
```
It should work, since we've specified the user to include the type of git host he prefers to use. BTW, bit bucket does support git.

abhsag24 replied:
i meant in the url

nkprince007 replied:
Yeah, it would... :P any suggestions on how to fix it @abhsag24 ?

nkprince007 replied:
So, how about just asking the user to specify the host he/she uses instead of VCS?

abhsag24 replied:
@nkprince007 that is better

nkprince007 commented:
@Mixih, but we still could modify them accordingly with .coafile

abhsag24 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -50,7 +52,9 @@ def get_metadata(cls):
 
-    def run(self, allow_empty_commit_message: bool = False, **kwargs):
+    def run(self,
+            allow_empty_commit_message: bool = False,
```
why?

abhsag24 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -183,6 +189,9 @@ def check_body(self, body,
         :param body_line_length:    The maximum line-length of the body. The
                                     newline character at each line end does not
                                     count to the length.
+        :param host_used:           The version control system used.
```
shouldnt this be the host website used or something?

nkprince007 replied:
@abhsag24 Will update the terms used. :+1:

nkprince007 commented:
@Mixih but we could still use `coala -b=GitCommitBear -S commit.host_website=bitbucket` and it still works, right?
Or we could just add the value to `.coafile` to make it work, right?

Mixih commented:
Yeah, I missread the code, it should work...

nkprince007 commented:
Ok, bro. I'll make the changes in the noon, once I finish my exam.

jayvdb commented on .coafile:
```diff
@@ -29,6 +29,8 @@ bears = SpaceConsistencyBear
 [commit]
 bears = GitCommitBear
+host_used = github
+contains_url = True
```
This is going to cause gitmate to reject any "multiple commits PR" where each commit does not have an issue linked.  i.e. it is going to reject quite a few of our commits. e.g.  coala/coala@ea48d66b9 coala/coala@49f8884ff coala/coala@49f441bf2 .  coala/coala@39b0a170 .
Are we ready for this?

nkprince007 replied:
I'm reverting the `.coafile` back to normal anyway @jayvdb 
But, do suggest me how to differentiate between issue related commits and non issue related ones?

nkprince007 replied:
Do I combine url presence with the presence of git keywords as suggested [here](https://help.github.com/articles/closing-issues-via-commit-messages/)? @jayvdb

nkprince007 commented:
I've added the tests and adjusted the bear to check for presence of url only when a git keyword to close issues is present in the commit message body. This won't cause any trouble with `git-mate`

jayvdb commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -175,6 +177,7 @@ def check_imperative(self, paragraph):
     def check_body(self, body,
                    body_line_length: int=72,
                    force_body: bool=False,
+                   host_site: str='',
```
default should be `None`.

jayvdb commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -197,6 +202,34 @@ def check_body(self, body,
 
+        # Git keywords for closing issues
+        # https://help.github.com/articles/closing-issues-via-commit-messages/
+        git_keys = """close closes closed
```
store this in a `set` as a module or class attribute

jayvdb replied:
Also note that these are *GitHub* key words.  git does not know about them.

1. investigate whether the gitlab has the same keywords
2. rename the variable to `github_...` if the set is unique to github

jayvdb commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -197,6 +202,34 @@ def check_body(self, body,
+                netloc = pieces.netloc
+                scheme = pieces.scheme
+                path = pieces.path
+                try:
```
what is this?
It works hard, but has no result.

jayvdb commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -197,6 +202,34 @@ def check_body(self, body,
+                    assert all([scheme, netloc, path])
+                except AssertionError:
+                    pass
+                if netloc.find(host_site) != -1 and path.find('issues') != -1:
```
At least use `/issues/`.

But also please check that the part after `/issues/` is a number.

Mixih commented:
disable your circle building. Master circle has been fixed now. After that commit -a --amend with no modifications and force push. This should fix the builds.

sils commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -175,6 +180,7 @@ def check_imperative(self, paragraph):
     def check_body(self, body,
                    body_line_length: int=72,
                    force_body: bool=False,
+                   host_site: str=None,
```
I thikn that'd be way better

sils commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -183,6 +189,8 @@ def check_body(self, body,
                                     newline character at each line end does not
                                     count to the length.
+        :param host_site:           The website used for hosting the git
+                                    repository.
```
the docs dont make clear that this will check for full URLs

sils commented:
marked WIP, let's work out the general issues first, then get into the details :)

abhsag24 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -21,6 +23,9 @@ class GitCommitBear(GlobalBear):
     LICENSE = 'AGPL-3.0'
     ASCIINEMA_URL = 'https://asciinema.org/a/e146c9739ojhr8396wedsvf0d'
     CAN_DETECT = {'Formatting'}
+    GITHUB_KEYS = {'close', 'closes', 'closed', 'resolve',
```
why not make this configurable as well? then we can make it for everyone?

abhsag24 replied:
@sils @jayvdb if you could confirm that.

abhsag24 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -197,6 +207,36 @@ def check_body(self, body,
+            body_string = '\n'.join(body).lower()
+            keys = set()
+
+            if 'github' in host_site.lower():
```
making it configurable will also save you these :P

abhsag24 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -197,6 +207,36 @@ def check_body(self, body,
+                keys = self.GITLAB_KEYS
+
+            if any(key in body_string for key in keys):
+                urls = re.findall(r'(https?://\S+)', body_string)
```
what if the commit structure is like this
```
<tag>: shortlog

as we can see in the url: <https://......>

Fixes #109
```
your bear would search for the url in the body which is not related to `Fixes` and would yield a Result right? 
just check it after the fixes?

nkprince007 replied:
Yeah. @jayvdb suggested use of full urls. But I'm working on making it configurable as well.

abhsag24 commented:
marking WIP since the feature doesn't check all cases

dracarys09 commented on tests/vcs/git/GitCommitBearTest.py:
```diff
@@ -267,6 +269,108 @@ def test_body_checks(self):
+        self.assertEqual(self.run_uut(issue_checks=True),
+                         ['No issue related URL found in '
+                          'the body at HEAD commit. '
+                          'Please add one.'])
```
The issue related url is present but it is not a valid issue related url... The message should be changed accordingly.

nkprince007 replied:
Ok @dracarys09

dracarys09 commented on tests/vcs/git/GitCommitBearTest.py:
```diff
@@ -267,6 +269,108 @@ def test_body_checks(self):
+                                      keyword_regex='Fix\s+#[0-9]+\s*'),
+                         ['No match found for the regular '
+                          'expression provided: Fix\s+#[0-9]+\s*'])
+
```
Your regex will also match lines like : "Fix #007"
However I think this is not a correct behaviour because issues I think never start from '0'
You can change [0-9] to [1-9] and add a check for this this type of urls as well.

nkprince007 replied:
[1-9] ? but that disables 0 even in numbers like 70.

Although anyway that was just a test, right?
It depends on the user to correctly write the regex.

dracarys09 replied:
I agree that it was just a test. I meant to say that you can add one more test to check that the user does not enter something like 005 etc.  btw about the regex my bad... I edited the regex... pls check now. 😄

nkprince007 replied:
Okay, I'll change the test, but you know that regex is given user freedom, right? Why do we need to check it again when user specifies that himself?

dracarys09 replied:
The point is no issues will start like 01 or 005 etc... so if 'by mistake' a user uses ```Fixes #07``` then we should be able to convey this to the user.

jayvdb commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -81,6 +98,16 @@ def run(self, allow_empty_commit_message: bool = False, **kwargs):
+        stdout, _ = run_shell_command('git config --get-regex url')
+
+        stdout = stdout.splitlines()
+        stdout = [url.split()[-1] for url in stdout]
```
this should have an appropriate variable name, like `remotes`.

nkprince007 replied:
Sure @jayvdb

jayvdb commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -81,6 +98,16 @@ def run(self, allow_empty_commit_message: bool = False, **kwargs):
             stdout[1:],
             **self.get_body_checks_metadata().filter_parameters(kwargs))
 
+        stdout, _ = run_shell_command('git config --get-regex url')
```
This should be in `check_issue` , and it is wasted computation if none of the relevant options are enabled.

nkprince007 replied:
I always forget the minute details. 😞

jayvdb replied:
ah, my bad, `--get-all` doesnt accept wildcards.

jayvdb commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -204,3 +224,71 @@ def check_body(self, body,
+                yield Result(self, 'No match found for the regular '
+                                   'expression provided: %s' % keyword_regex)
+                return
+
```
add `return` here also, for when `re.search(..)` was `True`.

Then you do not need the `else:` below.
Removing the `else` means one less indent, and also the Cyclomatic Complexity should be lower

jayvdb commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -204,3 +224,71 @@ def check_body(self, body,
+                    netloc = urlparse(url)[1]
+                    host = netloc.split('.')[0]
+                else:
+                    netloc = url[url.find('@')+1: url.find(':')]
```
no space after the colon.
run your code through pycodestyle before submitting.

jayvdb commented:
I find the two settings `issue_checks` and `keyword_regex` to be confusing, and not well designed for when we want to add new settings for issue checks.

There are some projects which want the literal `https://github.com/coala/coala-bears/issues/123`.  coala wants this.

Other hosts, such as bitbucket, and other projects that want the literal `#123` and do not want a URL.
That is https://github.com/coala/coala-bears/issues/1134

IMO we need an extra config settings in our .coafile ,  something like `require_issue_urls = True` , which rejects the literal `#123` even though GitHub and GitLab recognise it.

i.e. you need to partially design https://github.com/coala/coala-bears/issues/1134 in your pull request , though of course you do not need to build it all, especially not building tests for bitbucket repositories.

jayvdb commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -204,3 +224,81 @@ def check_body(self, body,
+        :param body:                Body of the commit message of HEAD.
+        :param detect_host:         Whether to automatically detect host
+                                    from git config or not.
+        :param require_url:         Whether the commit message requires
```
As there are only two values, this should be a `bool`.
Default disabled, as urls are not mandatory on github/gitlab.  This means `.coafile` in coala repos need to enable this flag.

nkprince007 replied:
Ok, bro. But what about short urls?

jayvdb replied:
sorry, what are short urls? https://git.io/ ??

jayvdb replied:
@nkprince007 what are short urls?

nkprince007 replied:
^ corrected what i meant! :stuck_out_tongue: I was asking about #<issue>

jayvdb commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -204,3 +224,81 @@ def check_body(self, body,
+                                    'short_ref' confines commit to have
+                                    shorter issue references like #123, etc.
+        :param keyword_regexp:      If provided, matches the keywords
+                                    with the regex. This gains precedence
```
gains->takes

jayvdb commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -204,3 +224,81 @@ def check_body(self, body,
+                                    shorter issue references like #123, etc.
+        :param keyword_regexp:      If provided, matches the keywords
+                                    with the regex. This gains precedence
+                                    over auto_detect_host and require_url.
```
"auto_detect_host"?  an old name.

nkprince007 replied:
Yeah. Thanks :smile:

Makman2 replied:
Docstring is not correctly aligned, either do
```
:param xyz: Hello
:param x:    Hello2
```
or (what I recommend as it eases maintenance of long docstrings):
```
:param xyz:
    Hello
:param x:
    Hello2
```

nkprince007 replied:
The second one looks more welcoming. :heart:

Mixih commented:
WIP as features are under discussion

Mixih commented:
hmm, what was I thinking? marking WIP stops discussion.. oh well, undoing.

nkprince007 commented:
Sure @mixih. I wasn't gonna touch it either way, till we get to a conclusion.

Makman2 commented on .coafile:
```diff
@@ -29,6 +29,7 @@ bears = SpaceConsistencyBear
 
 [commit]
 bears = GitCommitBear
+issue_checks = True
```
1. I think this should go into a separate commit
2. you don't define `issue_checks` as a settings, instead you have
    ```
    detect_host: bool=False,
    require_url: str=None,
    keyword_regexp: str=None
    ```
    So either unify them to an `issue_checks` setting, otherwise you need to use those settings^^

nkprince007 replied:
Yeah, sure. Btw, awesome work on this bear @Makman2

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -204,3 +224,81 @@ def check_body(self, body,
+                host = netloc.split('.')[0]
+                continue
+            netloc = urlparse(url)[1]
+            host = netloc.split('.')[0]
```
you override `host` always, this won't work really when having two URLs in the message, does it?

nkprince007 replied:
@Makman2, this is the part which checks `.git/config` for remote urls for the host. In all use case scenarios, we generally have the same host, so, I don't think it really matters. In fact I think it's better to go by the first one without iterating. 
Thanks for the review. :smile:

Makman2 replied:
but not in all cases, sometimes I use two remotes at once (especially for forks). I think the remotes-check is not necessary, just try to find github or gitlab urls (so all hosts you support) inside the commit message, regardless whether the remote is set to the actual host.

nkprince007 replied:
So, you're suggesting to pick the message from commit body and identify the host, then check for presence of the site's keywords? That looks awesome 😮

Makman2 replied:
yes :)
To extract links, see `InvalidLinkBear`. It has a good function to extract links, it's also quite battle tested, you should use it somehow (maybe we should put it into the core repository).

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -204,3 +224,81 @@ def check_body(self, body,
+        for url in remotes:
+            netloc = urlparse(url)[2]
+            if 'git@' in url:
+                netloc = url[url.find('@')+1:url.find(':')]
```
what about using a regex?
```
import re
re.find('@(.*):')
```
your current find-approach would work wrong when having this:
```
:xyz-git@123:
```

nkprince007 replied:
Yeah, sure. I'll replace it with a regexp.

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -204,3 +224,81 @@ def check_body(self, body,
+            host = netloc.split('.')[0]
+
+        if host not in self.SUPPORTED_HOSTS:
+            return
```
okay you check up with the configured remotes of the repo, you sure that this is needed? If you mirror a github repo on bitbucket (not sure this is possible, but we just assume it is), then your bear wouldn't check the issue references though they point to github^^

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -204,3 +224,81 @@ def check_body(self, body,
+
+        matches = re.findall(r'https?://\S+', body)
+
+        for match in matches:
```
inline stmt from above:
```
for match in re.findall(...):
```

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -204,3 +224,81 @@ def check_body(self, body,
+
+        if not good_ref:
+            yield Result(self, 'No issue related reference found in the '
+                               'body at HEAD commit. Please add one.')
```
duplicate error message, maybe you can use variable?

jayvdb commented:
@Makman2, could you instead comment on the design review at https://github.com/coala/coala-bears/issues/1112#issuecomment-267953755

jayvdb commented:
unack 5704451

jayvdb commented:
I've changed the issue to be pending review, so we get feedback in the correct spot.

jayvdb commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +223,88 @@ def check_body(self, body,
+            body = body.splitlines()[-1]
+
+        if not any(key in body for key in keys):
+            yield Result(self, 'No issue related voodoo found in the '
```
"voodoo"

nkprince007 replied:
"Commit doesn't seem to relate to any issue" Would that be good @jayvdb ?

nkprince007 commented:
Roger that @jayvdb. Moving onto writing unit tests. :smiley_cat:

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +223,80 @@ def check_body(self, body,
+            precedence over body_close_issue.
+        :param body_close_issue:
+            Whether to automatically detect host from git config or not.
+            If detected checks for the presence of issue close voodoo
```
what is voodoo :O I would be more precise, we are engineers, not magicians :)

nkprince007 replied:
:smile: docs haven't been corrected, I see.

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +223,80 @@ def check_body(self, body,
+            If provided, matches the keywords with the regex. This takes
+            precedence over body_close_issue.
+        :param body_close_issue:
+            Whether to automatically detect host from git config or not.
```
`body_close_issue` -> Whether to automatically detect host from git config or not

Those do not really fit together imo, maybe something like `use_git_config_remotes` or so?

nkprince007 replied:
But, the decision was based on comments in #1112 related to variable naming.

Makman2 replied:
I think you should just leave this sentence out and take the following one, so
```
Whether to check for presence of short references...
```

Makman2 replied:
Btw don't you support all issue references with this flag, also full URLs?

nkprince007 replied:
It's more like, if this flag is enabled, it checks for any reference, be it short or full URL. But, by default, it checks for short references, if other flags are enabled, works otherwise.

Makman2 replied:
> It's more like, if this flag is enabled, it checks for any reference, be it short or full URL.

okay.

> But, by default, it checks for short references

huh? I'm getting confused...

nkprince007 replied:
^^corrected.

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +223,80 @@ def check_body(self, body,
+            checks for presence of short references like 'closes #213', etc.
+        :param body_close_issue_url:
+            Checks the presence of issue close reference with a full URL
+            related to some issue. Works along with body_close_issue.
```
I would rename to `body_needs_close_issue_url`^^

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +223,80 @@ def check_body(self, body,
+        :param body_close_issue_url:
+            Checks the presence of issue close reference with a full URL
+            related to some issue. Works along with body_close_issue.
+        :param body_close_issue_last_line:
```
maybe also `body_close_issue_on_last_line`?

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +223,80 @@ def check_body(self, body,
+            Whether to automatically detect host from git config or not.
+            If detected checks for the presence of issue close voodoo
+            within the commit body. By default, if none of
+            body_close_issue_url and body_close_issue_last_line are enabled,
```
enclose those variables in double backticks, so they get rendered as code^^

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +223,80 @@ def check_body(self, body,
+            If detected checks for the presence of issue close voodoo
+            within the commit body. By default, if none of
+            body_close_issue_url and body_close_issue_last_line are enabled,
+            checks for presence of short references like 'closes #213', etc.
```
I would also use double backticks instead of `'`

Makman2 replied:
But yes, which short references are supported? You could add another paragraph here explaining them all, otherwise a user can't know^^

nkprince007 replied:
Maybe add a link to GitHub and GitLab docs for the auto closing keywords?

Makman2 replied:
yes good idea :+1:

nkprince007 replied:
How to put in a hyperlink in docs? this way `[test](http://example.com)`?

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +223,80 @@ def check_body(self, body,
+            Body of the commit message of HEAD.
+        :param body_regex:
+            If provided, matches the keywords with the regex. This takes
+            precedence over body_close_issue.
```
Use double backticks for `body_close_issue`

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +223,80 @@ def check_body(self, body,
+
+        :param body:
+            Body of the commit message of HEAD.
+        :param body_regex:
```
So this setting checks the body with a regex, right? Imo this belongs to `check_body`. Also I think if you enable it, it doesn't have to take precedence. For example If I just want that a body message starts with some kind of annotation, like this:
```
JIRA TICKET: [FLINK-1324]

blablablabla body message blabla
```
and my regex would be this one: `JIRA TICKET: \[FLINK-\d+](\n\n.*)?`

Maybe I still want to enforce the URL checks, as my regex is not specific enough + I don't want to reimplement that what you do with this PR inside my regex.

nkprince007 replied:
:smile: Sure, sounds like a welcome change.

Makman2 replied:
If you do this, please use another commit :+1:

nkprince007 replied:
Just a small change, oneliner. Does it need a new commit?

Makman2 replied:
yes, as it's an unrelated change^^

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +223,80 @@ def check_body(self, body,
+            within the commit body. By default, if none of
+            body_close_issue_url and body_close_issue_last_line are enabled,
+            checks for presence of short references like 'closes #213', etc.
+        :param body_close_issue_url:
```
As you explicitly check for a full url, I would name the setting accordingly: `body_close_issue_full_url`

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +223,80 @@ def check_body(self, body,
+            related to some issue. Works along with body_close_issue.
+        :param body_close_issue_last_line:
+            When enabled, checks for issue close reference presence on the
+            last line of the commit body. Works along with body_close_issue.
```
double backticks for `body_close_issue`

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +223,80 @@ def check_body(self, body,
+            if not re.search(body_regex, body):
+                yield Result(self, 'No match found for the regular '
+                                   'expression provided: %s' % body_regex)
+            return
```
please a newline here

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +223,80 @@ def check_body(self, body,
+                                   'expression provided: %s' % body_regex)
+            return
+        if not body_close_issue:
+            return
```
and here too a newline

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +223,80 @@ def check_body(self, body,
+            return
+        if not body_close_issue:
+            return
+        remotes, _ = run_shell_command('git config --get-regex url')
```
what if git fails as it's not a git directory? stdout would be empty and stderr would contain something, you should check that^^

Makman2 replied:
then I would directly suggest you call the variables `git_stdout, git_stderr` (or just `stdout, stderr`) instead of `remotes, _`

nkprince007 replied:
That's already handled [here](https://github.com/coala/coala-bears/blob/master/bears/vcs/git/GitCommitBear.py#L63-L67) :smile:

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +223,80 @@ def check_body(self, body,
+                netloc = re.findall(r'@(\S+):', url)[0]
+            else:
+                netloc = urlparse(url)[1]
+            host = netloc.split('.')[0]
```
newline please

Makman2 replied:
\+ I think you could this into an own function, like `get_hosts_from_git_remotes` or so

Makman2 replied:
and what about multiple remotes now?

Makman2 replied:
and what about multiple remotes now?

Makman2 replied:
oh yes, you only get the first element you find, this cries for an iterator and then you should rephrase to this:
```
try:
    host = next(get_hosts_from_git_remotes(remotes)):
except StopIteration:
    return  # <- No remotes registered
```

Makman2 replied:
or you can use a default value:
```
host = next(..., None)
```

nkprince007 replied:
No idea how to tackle that. But, considering this PR, how would we describe something like this.
```
Shortlog: Adjusting the sails

The ship now flies better.
Fixes https://github.com/jedi/swordfight/issues/1212
```
If we have two remotes `github` and `gitlab`. It doesn't make sense to me how to understand which one to match it against.

Makman2 replied:
Alright, be sure to file an issue for this after merge :)

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +223,80 @@ def check_body(self, body,
+                netloc = urlparse(url)[1]
+            host = netloc.split('.')[0]
+        if host not in self.SUPPORTED_HOSTS:
+            return
```
here too

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +223,80 @@ def check_body(self, body,
+        if body_close_issue_last_line:
+            body = body.splitlines()[-1]
+            result = ('No {} reference found in the body at HEAD commit in '
+                      'the last line.')
```
rather use an else clause instead of this override pattern:
```
if body_close_issue_last_line:
    body = ...
    result = ...
else:
    result = 'No {} reference found in the body at HEAD commit.'
```

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +223,80 @@ def check_body(self, body,
+        if host not in self.SUPPORTED_HOSTS:
+            return
+        keys = self.SUPPORTED_HOSTS[host]
+        body = body.lower()
```
For performance reasons (not highly relevant, though still an improvement) I would suggest you move this below `if body_close_issue_last_line:`

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +223,80 @@ def check_body(self, body,
+                      'the last line.')
+        if not any(key in body for key in keys):
+            yield Result(self, result.format('issue'))
+            return
```
please use newlines after if's and for's ... :D

nkprince007 replied:
Yeah, actually I removed them only trying to reduce the size of the PR. :smile: 
I'll add them back, looks like I've spoiled the readability.

Makman2 replied:
that's a bad way of decreasing PR size :)

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +223,80 @@ def check_body(self, body,
+            if not re.search('#[1-9][0-9]*', body):
+                yield Result(self, result.format('short issue'))
+            return
+        for match in re.findall(r'https?://\S+', body):
```
what about `http`?

nkprince007 replied:
dealt with already. `?` handles the existence of `s`

Makman2 replied:
oh oh yes yes sorry :+1:

jayvdb commented on tests/vcs/git/GitCommitBearTest.py:
```diff
@@ -267,6 +269,149 @@ def test_body_checks(self):
+                            body_close_issue=True,
+                            body_close_issue_full_url='full_ref'),
+                         ['Invalid issue number present in the body '
+                          'at HEAD commit: 1af121'])
```
this can be confusing. Is 1af121 a git sha?

nkprince007 replied:
No, just some random combination of ASCII characters. It actually should be a number. Else, it'd be rejected.

Makman2 replied:
yes please use 7 chars, this is a valid short-form git sha and does not confuse :)

nkprince007 replied:
Ok. @Makman2 :+1:

jayvdb commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +239,85 @@ def check_body(self, body,
                                'Commit body lines should not exceed {} '
                                'characters.'.format(body_line_length))
+
+    def check_issue(self, body,
```
Please check the values match the spec.

nkprince007 replied:
?? I don't get it @jayvdb

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -66,6 +98,7 @@ def run(self, allow_empty_commit_message: bool = False, **kwargs):
             self.err('git:', repr(stderr))
             return
 
+        body = stdout.rstrip('\n')
```
why do you do this?

nkprince007 replied:
To keep it as a single string, so that regular expressions could be searched for easily.

Makman2 replied:
actually this is not the commit body, it also contains the shortlog. Also you are doing double work with the stmt below:
```
stdout =s tdout.ristrip('\n').splitlines()
```
For consistency I would suggest you also introduce a `shortlog` variable:
```
shortlog = ...
body = stdout[:stdout.find('\n')] # <-- this is not idiot proof, you need to check empty commit messages,
# single line commit message and multiline commit message for proper results.
...
yield from self.check_shortlog(shortlog, ...)
yield from self.check_body(body.splitlines())
yield from self.check_issue(body)
```

Makman2 replied:
introduction of a `shortlog` and `body` variable could be also done in a separate commit :+1:

nkprince007 replied:
I assume that
```
body = stdout[:stdout.find('\n')] 
```
is a safe move.
Because that's essentially what's being done [here](https://github.com/coala/coala-bears/blob/master/bears/vcs/git/GitCommitBear.py#L76-L81)

nkprince007 replied:
But there's also the case where we have shortlog only, where this inherently fails.

nkprince007 replied:
I think it's better leaving it this way.
```
stdout = stdout.rstrip('\n').splitlines()
body = '\n'.join(stdout[1:])
```

Makman2 replied:
yeah joining again shouldn't be a performance impact^^

though this works:
```
pos = stdout.find('\n')
shortlog = stdout if pos == -1 else stdout[0:pos]
body = '' if pos == -1 else stdout[pos+1:]
```

nkprince007 replied:
Looks good :smile:

Makman2 replied:
yes a separate commit makes it way easier to review. In fact you standardize and improve shortlog and body extraction :)

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +239,85 @@ def check_body(self, body,
                                'characters.'.format(body_line_length))
+
+    def check_issue(self, body,
+                    body_regex: str=None,
```
Didn't we agree to move this param to `check_body`?

nkprince007 replied:
Oh, I was thinking of it the other way around. :worried:

nkprince007 replied:
Want me to move it? @Makman2

Makman2 replied:
yes please :+1:

Makman2 replied:
and into a separate commit

nkprince007 replied:
But this would affect the use of `body.splitlines()` :worried: Would cause me to join the lines again.

Makman2 replied:
don't get you. The `body_regex` feature is something separate, so imo it belongs to an own commit. If it needs adjustments elsewhere, then yes, do them :)

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +239,85 @@ def check_body(self, body,
+            if not re.search(body_regex, body):
+                yield Result(self, 'No match found for the regular '
+                                   'expression provided: %s' % body_regex)
+            return
```
No return, as I pointed out we want the issue check still be running though a regex was selected for the body^^

nkprince007 replied:
change already housed in the second commit :P

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -41,13 +50,36 @@ def get_body_checks_metadata(cls):
+            else:
+                netloc = urlparse(url)[1]
+            host = netloc.split('.')[0]
+        return host
```
fails with an exception when no remotes are set, you should set `host` to `None` before^^

nkprince007 replied:
Yeah :worried:

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -41,13 +50,36 @@ def get_body_checks_metadata(cls):
+                netloc = re.findall(r'@(\S+):', url)[0]
+            else:
+                netloc = urlparse(url)[1]
+            host = netloc.split('.')[0]
```
you only use one remote. you don't need to iterate. Just get the first element and process it (if it does exist)^^

nkprince007 replied:
Yeah, sure. @Makman2

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -99,19 +99,21 @@ def run(self, allow_empty_commit_message: bool = False, **kwargs):
             return
 
-        body = stdout.rstrip('\n')
-        stdout = stdout.rstrip('\n').splitlines()
```
this change has to be removed from this commit^^

Makman2 replied:
nvm...

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -66,6 +99,7 @@ def run(self, allow_empty_commit_message: bool = False, **kwargs):
             self.err('git:', repr(stderr))
             return
 
+        body = stdout.rstrip('\n')
```
you should move the commit introducing the `shortlog` and `body` variables before this commit and use them accordingly here^^

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -41,13 +50,37 @@ def get_body_checks_metadata(cls):
+    @staticmethod
+    def get_host_from_remotes():
+        """
+        Retrieve the host from supplied git remotes.
```
I think you should point out in the docs that you return the first remote that appears in `git remotes` :)

nkprince007 replied:
Ok :smile:

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +240,85 @@ def check_body(self, body,
+        :param body:
+            Body of the commit message of HEAD.
+        :param body_regex:
+            If provided, matches the keywords with the regex.
```
does not exist in this commit :)

Makman2 replied:
but I think you wanted to move that into an own commit?

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +240,85 @@ def check_body(self, body,
+            of short references like ``closes #213``, etc. Otherwise behaves
+            according to other chosen flags.
+            More on keywords follows.
+            [GitHub keywords](https://help.github.com/articles/closing-issues-via-commit-messages/)
```
hm this breaks our line-length limit, though I like those links... Is it possible to break them up?

nkprince007 replied:
But links are already configured to break line length limits, right? As the coafile goes?

Makman2 replied:
oh seems like, nvm then :)

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +240,85 @@ def check_body(self, body,
+            commit messages. Checks for matching keywords in the commit body.
+            By default, if none of  ``body_close_issue_full_url`` and
+            ``body_close_issue_on_last_line`` are enabled, checks for presence
+            of short references like ``closes #213``, etc. Otherwise behaves
```
etc?

nkprince007 replied:
Isn't it the correct way? :thinking:

Makman2 replied:
in this case I don't know what you mean with "etc" :) I would omit it :+1:

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +240,85 @@ def check_body(self, body,
+            By default, if none of  ``body_close_issue_full_url`` and
+            ``body_close_issue_on_last_line`` are enabled, checks for presence
+            of short references like ``closes #213``, etc. Otherwise behaves
+            according to other chosen flags.
```
So you can either check for short refs or long refs, but not both in the same time?

nkprince007 replied:
Nope.

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -113,7 +113,7 @@ def run(self, allow_empty_commit_message: bool = False, **kwargs):
             shortlog,
```
ahh...
please don't implement that in a commit before, and then you move it. Make it right from the beginning, implement it directly in `self.check_body` :+1:

nkprince007 replied:
Ok. :blush:

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -213,23 +213,34 @@ def check_imperative(self, paragraph):
             return
 
+        if body_regex:
+            if not re.search(body_regex, body):
```
not `re.match`?

Makman2 replied:
oh and another thing: This function implicitly checks for a newline between shortlog and body. Maybe you should strip the first `\n` away or adjust the regex to match it implicitly.
\+ the regex check should then happen after the "newline"-check^^

nkprince007 replied:
Ok.. :smile:

nkprince007 replied:
not `re.match()` as it only searches for the beginning of string. More about this [here](https://docs.python.org/3/library/re.html#search-vs-match)

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -213,23 +213,34 @@ def check_imperative(self, paragraph):
+        if body_regex:
+            if not re.search(body_regex, body):
+                yield Result(self, 'No match found for the regular '
+                                   'expression provided: %s' % body_regex)
```
-> No match found **in commit message** for the regular expression provided:

People could misunderstand where this message comes from^^

nkprince007 replied:
Yeah definitely :laughing:

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -213,23 +213,34 @@ def check_imperative(self, paragraph):
+            if not re.search(body_regex, body):
+                yield Result(self, 'No match found for the regular '
+                                   'expression provided: %s' % body_regex)
+                return
```
I think you don't need to stop analysis here :)

Makman2 commented:
ack 1332cf4

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -3,6 +3,8 @@
 import shutil
```
commit message: The part with 

> If keyword regex is chosen, it verifies
> the presence of matching expression.

doesn't apply any more :)

nkprince007 replied:
?? @Makman2 :worried:

nkprince007 replied:
?? @Makman2 :worried:

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +240,76 @@ def check_body(self, body,
+        if host not in self.SUPPORTED_HOSTS:
+            return
+
+        keys = self.SUPPORTED_HOSTS[host]
```
hm maybe call it `keywords`? `keys` confuses me slightly :)

nkprince007 replied:
Ok.

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +240,76 @@ def check_body(self, body,
+            return
+
+        keys = self.SUPPORTED_HOSTS[host]
+        good_ref = False
```
as you use those variables below the coming if, you should move them below too :)

nkprince007 replied:
Sure.

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +240,76 @@ def check_body(self, body,
+            result = 'No {} reference found in the body at HEAD commit.'
+        else:
+            body = body.splitlines()[-1]
+            result = ('No {} reference found in the body at HEAD commit in '
```
I would call it `result_message` :+1:

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +240,76 @@ def check_body(self, body,
+
+        for match in re.findall(r'https?://\S+', body):
+            netloc, path = urlparse(match)[1:3]
+            if not all([host in netloc, '/issues/' in path]):
```
I think you should just do
```
if not host in netloc or '/issues/' in path:
    continue
```
faster + better readable :)

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +240,76 @@ def check_body(self, body,
+                return
+
+        if not good_ref:
+            yield Result(self, result.format('issue'))
```
do you need this check again?

nkprince007 replied:
This is the only place where it is being checked. For eg. if there are keywords present but a url doesn't exist, it is detected here. :wink:

Makman2 commented on tests/vcs/git/GitCommitBearTest.py:
```diff
@@ -267,6 +269,127 @@ def test_body_checks(self):
+                        'Closes #1112')
+        self.assertEqual(self.run_uut(
+                            body_close_issue=True,
+                            body_close_issue_full_url=True), [])
```
indentation: 4 spaces instead of 3 relative to opening statement :D
(same for all occurrences below)

aptrishu commented:
Other tests also need change according to your newly introduced parameter/s.
Also it needs some thinking when do we really want to show `No short issue reference found in the body at HEAD commit.`. We'll discuss it on channel.

Makman2 commented on tests/vcs/git/GitCommitBearTest.py:
```diff
@@ -269,6 +269,24 @@ def test_body_checks(self):
+                        'Another line, blablablablablabla.\n'
+                        'Closes #01112')
+        self.assertEqual(self.run_uut(
+                            body_regex=r'Closes\s+#[1-9][0-9]*\s*'),
```
4 spaces indentation relative to opening expression/statement, same for the others above :)

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -213,28 +213,36 @@ def check_imperative(self, paragraph):
                                'HEAD commit. Please add one.')
             return
 
+        if body_regex and not re.search(body_regex, body):
```
so not `re.match` ? I think `re.match` is more appropriate here.

nkprince007 replied:
I think I already mentioned it before. re.match searches only in the beginning. It won't search the entire string.

Makman2 replied:
yes, but if I provide a body regex imo it should be matched against the whole body, not a part of it (maybe we should even use `re.fullmatch`, if 3.4 already supports it, but I believe it does). I would expect this behaviour:

```
this is the body of my commit message

TICKET XYZ-1234
```
if I provide a regex `TICKET XYZ-\d+` it shall fail as I only want to have those "TICKET" descriptions in the message. With `re.search` we simply can't check every kind of commit message^^

nkprince007 replied:
```
>>> import re
>>> re.search("TICKET XYZ-\d+", "this is the body of commit message\n\nTICKET XYZ-1234")
<_sre.SRE_Match object; span=(36, 51), match='TICKET XYZ-1234'>
>>> 
```
But it does return a match found.

nkprince007 replied:
And we are aiming at making it fail?

nkprince007 replied:
But what if I just want the issue URL to be present somewhere in the body, like this.
```
Shortlog: Add final fight with the Jedi

Adds light saber powers
Reduces damage taken
Adds double damage to punches and kicks

Closes https://github.com/starwars/jedi-war.git
```
Then I don't think we could use `re.match` or `re.fullmatch`.

Makman2 replied:
> And we are aiming at making it fail?

yes I would do so.

> But what if I just want the issue URL to be present somewhere in the body, like this.

just modify the regex:
```
(.|\n)*Closes http://(.*)/\d+(.|\n)*
```

nkprince007 replied:
But we couldn't use `re.match` to check the regex there. We're not just searching the beginning of body, right? To check the entire body string, we need `re.search`.

Makman2 replied:
Don't understand you. `re.search` searches for a match anywhere in the given string. `re.match` searches for match at the beginning of the given string. `re.fullmatch` searches for a match at the beginning of the given string, with the requirement that the complete text needs to be matched.

nkprince007 replied:
Yeah, exactly, wouldn't that cripple the capability of regex to be present only in the beginning? I mean to provide a better way, to search the entire body, while this change cripples it completely, searching only the start of the body.

Makman2 replied:
opening an issue for that after merge :)

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +240,76 @@ def check_body(self, body,
+            within the commit body by retrieving host information from git
+            configuration. GitHub and GitLab support auto closing issues with
+            commit messages. Checks for matching keywords in the commit body.
+            By default, if none of  ``body_close_issue_full_url`` and
```
redundant space after `of`

nkprince007 replied:
😊 ok.

Makman2 commented:
@aptrishu 
> Other tests also need change according to your newly introduced parameter/s.

Which ones? (Except the one that behaves weird, whyever it does)

> Also it needs some thinking when do we really want to show No short issue reference found in the body at HEAD commit.. We'll discuss it on channel.

Please post your results here, imo current behaviour is okay. Do you have something you want to change?

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -41,13 +50,39 @@ def get_body_checks_metadata(cls):
+        Retrieve the first host from the list of git remotes.
+        """
+        remotes, _ = run_shell_command(
+            'git config --get-regex \'^remote.*.url$\'')
```
rather use double quotes so you don't need to escape the single quotes:
```
"git config --get-regex '^remote.*.url$'"
```
btw if you want to improve performance, you can pre-split the command (up to you):
```
('git', 'config', '--get-regex', '^remote.*.url$')
```

nkprince007 replied:
`run_shell_command` has just one parameter. 😟

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +241,76 @@ def check_body(self, body,
+            according to other chosen flags.
+            More on keywords follows.
+            [GitHub](https://help.github.com/articles/closing-issues-via-commit-messages/)
+            [GitLab](https://docs.gitlab.com/ce/user/project/issues/automatic_issue_closing.html)
```
I think we should explicitly mention and make the user aware to configure remotes properly, as imo `[...] by retrieving host information from git configuration` is too vague.

But let's do this later, I'm filing an issue for that after merge.

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +241,76 @@ def check_body(self, body,
+                              'commit in the last line.')
+
+        keywords = self.SUPPORTED_HOSTS[host]
+        good_ref = False
```
`good_ref` is only used below, you should move it there^^

nkprince007 replied:
Yeah. :smile:

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +241,76 @@ def check_body(self, body,
+                continue
+            issue_num = path.split('/')[-1]
+            try:
+                issue_num = int(issue_num)
```
if you don't need `issue_num`, you don't need to reassign it :)
just do
```
int(issue_num)
```

nkprince007 replied:
I'm using it to show the result in line 323.

Makman2 replied:
I think you mean 313, and also if the `int` conversion fails the result won't be printed there anyway, you still get the string from before.

nkprince007 replied:
Oh, yeah.. I was thinking of it, incorrectly.

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +241,76 @@ def check_body(self, body,
+            except ValueError:
+                yield Result(self, 'Invalid issue number present in the '
+                                   'body at HEAD commit: %s' % issue_num)
+                return
```
maybe you should not return here, better report all wrong references.

Makman2 commented on tests/vcs/git/GitCommitBearTest.py:
```diff
@@ -267,6 +269,136 @@ def test_body_checks(self):
+        self.run_git_command('remote', 'add', 'test',
+                             'https://bitbucket.com/user/repo.git')
+
+        # Unsupported Host
```
you should maybe say in the comment before that BitBucket is the unsupported host :)

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +241,76 @@ def check_body(self, body,
+        if host not in self.SUPPORTED_HOSTS:
+            return
+
+        if not body_close_issue_on_last_line:
```
instead of inverting the condition, I would switch code of both cases and omit the `not`. You have code for both paths anyways :)

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +241,76 @@ def check_body(self, body,
+                yield Result(self, result_message.format('short issue'))
+            return
+
+        good_ref = False
```
maybe you should call it `has_good_ref` :)

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -20,6 +22,12 @@ class GitCommitBear(GlobalBear):
+        'github': (r'(?:[Cc]lose[sd]?|[Rr]esolve[sd]?'
+                   '|[Ff]ix(?:e[sd]?)?)\s+'),
+        'gitlab': (r'(?:[Cc]los(?:e[sd]?|ing)|[Rr]esolv(?:e[sd]?|ing)'
+                   '|[Ff]ix(?:e[sd]|ing)?)\s+')
```
sure you don't want to keep a simple list of keywords like before and join them dynamically with `|`?

nkprince007 replied:
I just wanna to stick to [GitLab docs](https://docs.gitlab.com/ee/user/project/issues/automatic_issue_closing.html)

Makman2 replied:
ah okay, cool
btw: I wouldn't enclose in `(?:...)`, this shall handle the regex inside the run-function later^^ it's redundant information

nkprince007 replied:
Ok.

nkprince007 replied:
I don't get it why it's behaving that way, but removal of `(?:...)` on GitLab specifically is causing an error. GitHub regex seems to work fine although.

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +240,71 @@ def check_body(self, body,
+
+        if body_close_issue_full_url:
+            result_message = result_message.format('full issue')
+            initial_check_regex = r'https?://{}\S+/issues/\w+'.format(host)
```
be sure to `re.escape` the host!

nkprince007 replied:
I don't think that's required. Since, it's actually captured from a remote, which is indeed a URL i.e. consists of ASCII characters.

Makman2 replied:
the problem if it accidentally has special regex chars, like `.` (take for example `http://mysite.com/page.html). It's general a good habit to escape text that you want to be full-text matched,

nkprince007 replied:
I understand the use case scenario, but it won't be a problem here, since the netloc of the URL as parsed via `urllib.parse.urlparse()` has already been processed from host to only contain only ASCII. But if you really wanna push me to use it, I'll replace it.

Makman2 replied:
do it :+1: :)
don't be lazy, make it conceptually right :+1:

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +240,71 @@ def check_body(self, body,
+        else:
+            result_message = result_message.format('issue')
+            initial_check_regex = r'#\w+'
+            final_check_regex = r'(#[1-9][0-9]*)'
```
1. I would rename the variables, initial-check and final-check are not really descriptive.
   maybe `issue_keyword_regex` and `issue_data_check_regex`? hopefully you have something better :)

Makman2 replied:
2. there's no 2. :P

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +240,71 @@ def check_body(self, body,
+            initial_check_regex = r'#\w+'
+            final_check_regex = r'(#[1-9][0-9]*)'
+
+        join_regex = r'({0}(?:\s*(?:,|and)\s*{0})*)'
```
--> join**t**_regex :D

nkprince007 replied:
:+1:

Makman2 replied:
I think you should make a variable for `(?:,|and)` to make code more readable:
```
concat_keywords = [',', 'and']  # <-- call it maybe differently
joint_regex = r'...(?:{})...'.format('|'.join(re.escape kw for kw in concat_keywords))
```

nkprince007 replied:
Just two keywords, do we know to go all the way along for that? 😟

Makman2 replied:
imo it improves readability. If see also later that different hosters support a different set of those keywords, we can easily move those also to a dict like `SUPPORTED_HOSTS`.

If you want, you can keep it as is, but if I see `,|and` occurring twice it's time to move those into a variable ;)

nkprince007 replied:
👍 Already seen it repeat twice. Introducing the variable

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +240,71 @@ def check_body(self, body,
+        join_regex = r'({0}(?:\s*(?:,|and)\s*{0})*)'
+        matches = re.findall(self.SUPPORTED_HOSTS[host]
+                             + join_regex.format(initial_check_regex), body)
+        matches = [re.findall(initial_check_regex, match) for match in matches]
```
I think you should really use `re.split` with `,|and`, as ideally `initial_check_regex` would be always something like `\w+` to match everything that follows after a keyword. After the split you can apply your `final_check_regex` stuff. This would also allow to easily combine short-links and full-link checking together, as you could just join these check-regexes (which we want to support later). Your current approach is imo not able to handle this without huge modification to all regexes / or to create new regexes while copying stuff from all existing ones together^^

nkprince007 replied:
I don't completely understand what you mean by 
> huge modification to all regexes
and 
> easily combine short-link and full-link checking together

nkprince007 replied:
I don't see much of a difference in using `re.split`
But, even if we used `re.split` for the stuff, we are still left with a list of lists of items, which again would need to be flattened into a list.

nkprince007 replied:
BTW, `initial_check_regex` is different for both cases, since I'm verifying URL right in itself in the first case for long URLs.
This is crucial for proper error reporting in issue numbers and wrong URLs.

Makman2 replied:
with `re.split` you get the list you want :) actually you don't need to flatten everything, just use a for-loop directly:
```
for issue_match in ...:
    for issue_id in re.split(issue_match, ...):
        pass
```

>> huge modification to all regexes
>
> Do you mean to say that we could combine the check for both of them with just one regular 
> expression?

yes :+1:

> BTW, initial_check_regex is different for both cases, since I'm verifying URL right in itself in the 
> first case for long URLs.
> This is crucial for proper error reporting in issue numbers and wrong URLs.

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +240,74 @@ def check_body(self, body,
+        joint_regex = r'({1}(?:\s*(?:{0})\s*{1})*)'.format(concat_regex,
+                                                           issue_keyword_regex)
+
+        matches = re.findall(self.SUPPORTED_KEYWORD_REGEX[host] + joint_regex,
```
assemble the host-keyword-regex before inside the `joint_regex` ;)

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +240,74 @@ def check_body(self, body,
+        concat_keywords = [',', 'and']
+        concat_regex = '|'.join(re.escape(kw) for kw in concat_keywords)
+        joint_regex = r'({1}(?:\s*(?:{0})\s*{1})*)'.format(concat_regex,
+                                                           issue_keyword_regex)
```
I think still that you shouldn't have different `issue_keyword_regex`es, because if I say: `body_close_issue_full_url=True` and have sth like this inside my message:
```
my super commit shortlog

body start.

Closes www.google.com/404
```
you won't match it at all. But we want to have a result saying that this is not a correct link (we want the http(s) prefix).

So I think `{0}` should be everything except whitespaces (or just `\w+` is also okay). For each match you find you process those elements matched by `\w+` and check if these are valid references (by splitting them up with `re.split`).

So the `issue_keyword_regex` should be used below on line 311, where you do
```
issue_keyword_regex.fullmatch(issue)
```

Makman2 replied:
oh btw you got a new test case with my example ;)

nkprince007 replied:
But using `\w+` would not match `https://github.com/u/r/issues/312`, right?

In fact, I tried many combinations like `\S+`, but couldn't figure out a way to match the entire URL.

Makman2 replied:
oh right, you could use `[^\s]*` instead then :+1:

Makman2 commented on bears/vcs/git/GitCommitBear.py:
```diff
@@ -203,3 +240,74 @@ def check_body(self, body,
+            result_message = result_message.format('issue')
+            issue_keyword_regex = r'#\w+'
+
+        concat_keywords = [',', 'and']
```
you could move the keywords to class scope below `SUPPORTED_KEYWORD_REGEX`, so everything "data-related" is located there :)

nkprince007 commented:
Closed in favor of #1315.

This PR has too many comments and takes too long to load.
